modules/configuration/src/main/java/org/apache/ignite/internal/configuration/util/ConfigurationSerializationUtil.java

```diff
@@ -0,0 +1,406 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.configuration.util;
+
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Utility class to help serializing  deserializing configuration values - primitives, strings, or arrays of primitives or strings.
+ */
+public class ConfigurationSerializationUtil {
+    private static final byte BOOLEAN = 1;
+
+    private static final byte BYTE = 2;
+
+    private static final byte SHORT = 3;
+
+    private static final byte INT = 4;
+
+    private static final byte LONG = 5;
+
+    private static final byte CHAR = 6;
+
+    private static final byte FLOAT = 7;
+
+    private static final byte DOUBLE = 8;
+
+    private static final byte STRING = 9;
+
+    private static final byte ARRAY = (byte) 0x80;
+
+    private static final String[] EMPTY_STRING_ARRAY = new String[0];
+
+    /**
+     * Converts a configuration value into a byte array.
+     *
+     * @param value Configuration values.
+     * @return Serialized value.
+     */
+    public static byte[] toBytes(Object value) {
+        Objects.requireNonNull(value);
+
+        int header = header(value.getClass());
+
+        switch (header) {
+            case BOOLEAN:
+                return new byte[]{BOOLEAN, (byte) ((boolean) value ? 1 : 0)};
+
+            case BYTE:
+                return new byte[]{BYTE, (byte) value};
+
+            case SHORT:
+                return allocateBuffer(Short.BYTES + 1).put(SHORT).putShort((short) value).array();
+
+            case INT:
+                return allocateBuffer(Integer.BYTES + 1).put(INT).putInt((int) value).array();
+
+            case LONG:
+                return allocateBuffer(Long.BYTES + 1).put(LONG).putLong((long) value).array();
+
+            case CHAR:
+                return allocateBuffer(Character.BYTES + 1).put(CHAR).putChar((char) value).array();
+
+            case FLOAT:
+                return allocateBuffer(Float.BYTES + 1).put(FLOAT).putFloat((float) value).array();
+
+            case DOUBLE:
+                return allocateBuffer(Double.BYTES + 1).put(DOUBLE).putDouble((double) value).array();
+
+            case STRING: {
+                byte[] strBytes = ((String) value).getBytes(StandardCharsets.UTF_8);
+
+                return allocateBuffer(1 + strBytes.length).put(STRING).put(strBytes).array();
+            }
+
+            case BOOLEAN | ARRAY: {
+                boolean[] booleans = (boolean[]) value;
```

Comment:
Would it make sense to pack booleans as bits, not bytes?

Reply:
I wonder how many boolean arrays we will store. This is definitely a good idea. Should I implement it?

Reply:
Done

Reply:
Cool :) How about extracting the whole switch branch for boolean arrays to its own method, as it became pretty long now?